superset/db_engine_specs/superset.py

```diff
@@ -0,0 +1,435 @@
+
+from superset import db, security_manager, sql_parse
+from superset.db_engine_specs.sqlite import SqliteEngineSpec
+
+
+class SupersetEngineSpec(SqliteEngineSpec):
+    """
+    Internal engine for Superset
+
+    This DB engine spec is a meta-database. It uses the shillelagh library
+    to build a DB that can operate across different Superset databases.
+    """
+
+    engine = "superset"
+    engine_name = "Superset native database"
+    default_driver = ""
+    sqlalchemy_uri_placeholder = "superset://"
+
+    supports_file_upload = False
+
+
+# pylint: disable=abstract-method
+class SupersetAPSWDialect(APSWDialect):
+
+    """
+    A SQLAlchemy dialect for an internal Superset engine.
+
+    This dialect allows query to be executed across different Superset
+    databases. For example, to read data from the `birth_names` table in the
+    `examples` databases:
+
+        >>> engine = create_engine('superset://')
+        >>> conn = engine.connect()
+        >>> results = conn.execute('SELECT * FROM "superset.examples.birth_names"')
+
+    Queries can also join data across different Superset databases.
+
+    The dialect is built in top of the shillelagh library, leveraging SQLite to
+    create virtual tables on-the-fly proxying Superset tables. The
+    `SupersetShillelaghAdapter` adapter is responsible for returning data when a
+    Superset table is accessed.
+    """
+
+    name = "superset"
+
+    def create_connect_args(self, url: URL) -> Tuple[Tuple[()], Dict[str, Any]]:
+        return (
+            (),
+            {
+                "path": ":memory:",
+                "adapters": ["superset"],
+                "adapter_kwargs": {},
+                "safe": True,
+                "isolation_level": self.isolation_level,
+            },
+        )
+
+
+# pylint: disable=invalid-name
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def check_dml(method: F) -> F:
+    """
+    Decorator that prevents DML against databases where it's not allowed.
+    """
+
+    @wraps(method)
+    def wrapper(self: "SupersetShillelaghAdapter", *args: Any, **kwargs: Any) -> Any:
+        # pylint: disable=protected-access
+        if not self._allow_dml:
+            raise ProgrammingError(f'DML not enabled in database "{self.database}"')
+        return method(self, *args, **kwargs)
+
+    return cast(F, wrapper)
+
+
+def has_rowid(method: F) -> F:
+    """
+    Decorator that prevents updates/deletes on tables without a rowid.
+    """
+
+    @wraps(method)
+    def wrapper(self: "SupersetShillelaghAdapter", *args: Any, **kwargs: Any) -> Any:
+        # pylint: disable=protected-access
+        if not self._rowid:
+            raise ProgrammingError(
+                "Can only modify data in a table with a single, integer, primary key"
+            )
+        return method(self, *args, **kwargs)
+
+    return cast(F, wrapper)
+
+
+# pylint: disable=too-many-instance-attributes
+class SupersetShillelaghAdapter(Adapter):
```

Comment:
For clarity, would it make sense to place these in a separate package? It feels like this may be outside the scope of a typical db engine spec, and if we were to encapsulate this in a separate package (e.g. under `/extensions` or even just under `/db_engine_specs`, then it would be more clear that this is in fact not a db engine spec, but actually a Shillelagh adapter

Reply:
Yeah, I like the idea of an `extensions/`.

Reply:
Done!